### DISCUSSION TITLE
working code for GC9A01 Touch LCD module waveshare model.

### DISCUSSION THREAD

**Discussion by saturnhades** [moononournation/Arduino_GFX#419] [General]
Does anyone have working code for the display mentioned https://www.waveshare.com/1.28inch-touch-lcd.htm or the ESP32-S3 board of the same display https://www.waveshare.com/esp32-s3-touch-lcd-1.28.htm. I'm planning to buy it but due to some issues in past with working of the same non-touch version hesitant to do so. If any of you have code or working video with any libraries, please share. @moononournation please help

**Comment by AvinDeMar** [comment]
Hey, not sure if you already made a purchase but I have a similar setup and got the 1.28 inch display w/ touch to work. I have a slightly different model but the pins are the same (although named differently) and the boards are the same "GC9A01 driver and CST816S capacitive touch control chip Using SPI and I2C Communication respectively" and this line of wording is where I focused. 

In Adruino IDE, I used the TFT_eSPI library by 'Bodmer' for the display portion (note this does not support multiple displays natively), tested it, and then used the CST816S library by 'fbiego' for the touch. I'm using custom pin arrangements with the ESP32 but everything works. Hope this helps!

**User_Setup.h for the TFT_eSPI Library:**
#define USER_SETUP_INFO "User_Setup"
#define GC9A01_DRIVER
#define TFT_SDA_READ      // This option is for ESP32 ONLY, tested with ST7789 and GC9A01 display only (lucky us!)

// For ST7789, ST7735, ILI9163 and GC9A01 ONLY, define the pixel width and height in portrait orientation
#define TFT_WIDTH  240 // ST7789 240 x 240
#define TFT_HEIGHT 240 // GC9A01 240 x 240

// Pin definitions for TFT display (Change to your desired pin setup - My 1.28 display: MOSI = SDA, SCLK = SCL, BL = BLK, RST = RES, otherwise the same)
#define TFT_BL   5    // Backlight control pin (Some setups use -1 but I would like to control this during standby function)
#define TFT_CS   18   // Chip Select
#define TFT_DC   19   // Data Command control pin
#define TFT_RST  21   // Reset pin
#define TFT_MOSI 22   // SPI MOSI
#define TFT_SCLK 23   // SPI Clock

#define LOAD_GLCD   // Font 1. Original Adafruit 8 pixel font needs ~1820 bytes in FLASH
#define LOAD_FONT2  // Font 2. Small 16 pixel high font, needs ~3534 bytes in FLASH, 96 characters
#define LOAD_FONT4  // Font 4. Medium 26 pixel high font, needs ~5848 bytes in FLASH, 96 characters
#define LOAD_FONT6  // Font 6. Large 48 pixel font, needs ~2666 bytes in FLASH, only characters 1234567890:-.apm
#define LOAD_FONT7  // Font 7. 7 segment 48 pixel font, needs ~2438 bytes in FLASH, only characters 1234567890:-.
#define LOAD_FONT8  // Font 8. Large 75 pixel font needs ~3256 bytes in FLASH, only characters 1234567890:-.
//#define LOAD_FONT8N // Font 8. Alternative to Font 8 above, slightly narrower, so 3 digits fit a 160 pixel TFT
#define LOAD_GFXFF  // FreeFonts. Include access to the 48 Adafruit_GFX free fonts FF1 to FF48 and custom fonts
#define SMOOTH_FONT

#define SPI_FREQUENCY  27000000
#define SPI_READ_FREQUENCY  20000000

In my sketch I define the pins for the touch panel separately.

**Header of Sketch:**
#include <SPI.h>
#include <TFT_eSPI.h>  // Display library
#include <CST816S.h>   // Touch library
#include "Images.h"  // All Images if you are using an image file.

TFT_eSPI tft = TFT_eSPI();  // Create an instance of the display
CST816S touch(4, 16, 2, 15);  // Define the CST816S touch controller pins: SDA, SCL, RST, INT

**Reply by jscholtes** [reply]
got the ESP32-S3-Touch-LCD-1.28 running with:

#include <Arduino.h>
#include <Arduino_GFX_Library.h>

Arduino_DataBus *bus = new Arduino_ESP32SPI(8 /* DC */, 9 /* CS */, 10 /* SCK */, 11 /* MOSI */, 12 /* MISO */);
Arduino_GFX *gfx = new Arduino_GC9A01(bus, 14 /* RST */, 0 /* rotation */, true /* IPS */);


and following platform.io

[env:esp32s3usbotg]
platform = espressif32
board = esp32s3usbotg
framework = arduino
lib_deps = moononournation/GFX Library for Arduino@^1.5.4
